docs/DynamicCasting.md

```diff
@@ -0,0 +1,532 @@
+
+## AnyHashable (SE-0131)
+
+For casting purposes, `AnyHashable` behaves like a protocol type.
+
+## Existential/Protocol types
+
+Caveat:
+Protocols that have `associatedtype` properties or which make use of the `Self` typealias cannot be used as independent types.
+As such, the discussion below does not apply to them.
+
+Any Swift instance of a concrete type `T` can be cast to `P` iff `T` conforms to `P`.
+The result is a "protocol witness" instance that provides access only to those methods and properties defined on `P`.
+Other capabilities of the type `T` are not accessible from a `P` instance.
+
+The contents of a protocol witness can be accessed by casting to some other appropriate type:
+* For any protocol `P`, instance `t`, and type `U`, if `t is P`, then `t as? U == (t as! P) as? U`
+
+XXX TODO: The invariant above does not apply to AnyObject, AnyHashable.
+Does it suffice to explicitly exclude those two, or do other protocols share that behavior?  The alternative would seem to be to change the equality here into an implication.
+
+In addition to the protocol witness type, every Swift protocol `P` implicitly defines two other types:
+`P.Protocol` is the "protocol metatype", the type of `P.self`.
+`P.Type` is the "protocol existential metatype".
+These are described in more detail below.
+
+Regarding Protocol casts and Optionals
+
+When casting an Optional to a protocol type, the optional is preserved if possible.
+Given an instance `o` of type `Optional<T>` and a protocol `P`, the cast request `o as? P` will produce different results depending on whether `Optional<T>` directly conforms to `P`:
+
+* If `Optional<T>` conforms to `P`, then the result will be a protocol witness wrapping the `o` instance.  In this case, a subsequent cast to `Optional<T>` will restore the original instance.  In particular, this case will preserve `nil` instances.
+
+* If `Optional<T>` does not directly conform, then `o` will be unwrapped and the cast will be attempted with the contained object.  If `o == nil`, this will fail.  In the case of a nested optional `T???` this will result in fully unwrapping the inner non-optional.
+
+* If all of the above fail, then the cast will fail.
+
+For example, `Optional` conforms to `CustomDebugStringConvertible` but not to `CustomStringConvertible`.
+Casting an optional instance to the first of these protocols will result in an item whose `.debugDescription` will describe the optional instance.
+Casting an optional instance to the second will provide an instance whose `.description` property describes the inner non-Optional instance.
+
+## Array/Set/Dictionary Casts
+
+For Array, Set, or Dictionary types, you can use the casting operators to translate to another instance of the same outer container (Array, Set, or Dictionary respectively) with a different component type.
+Note that the following discussion applies only to these specific container types and does not currently generalize to other container types.
```

Comment:
*or to user-defined types. We get a lot of requests for this...

Reply:
Ah.  Good point.  I had intended "other container types" to cover that, but you're right it's not general enough.  I'll rephrase it.

Reply:
Rephrased.